plugin/edge/proxy.go

```diff
@@ -63,30 +116,65 @@ func (p *Proxy) Dial(proto string) (*dns.Conn, error) { return p.transport.Dial(
 // Yield returns the connection to the pool.
 func (p *Proxy) Yield(c *dns.Conn) { p.transport.Yield(c) }
 
-// Healthcheck kicks of a round of health checks for this proxy.
+// Healthcheck kicks off a round of health checks for this proxy.
 func (p *Proxy) Healthcheck() { p.probe.Do(p.Check) }
 
-// Down returns true if this proxy is down, i.e. has *more* fails than maxfails.
-func (p *Proxy) Down(maxfails uint32) bool {
-	if maxfails == 0 {
+// Down returns true if this proxy is down, i.e. has *more* fails than maxUpstreamFails.
+func (p *Proxy) Down(maxUpstreamFails uint32) bool {
+	if maxUpstreamFails == 0 {
 		return false
 	}
-
 	fails := atomic.LoadUint32(&p.fails)
-	return fails > maxfails
+	return fails > maxUpstreamFails
 }
 
-// close stops the health checking goroutine.
+// Stops the health checking and service pushing goroutines.
 func (p *Proxy) close() {
+	close(p.pushChan)
 	p.probe.Stop()
 	p.transport.Stop()
 }
 
-// start starts the proxy's healthchecking.
-func (p *Proxy) start(duration time.Duration) { p.probe.Start(duration) }
+// Starts the proxy's healthchecking.
+func (p *Proxy) start(healthCheckDuration time.Duration) {
+	p.probe.Start(healthCheckDuration)
+}
 
-const (
-	dialTimeout = 4 * time.Second
-	timeout     = 2 * time.Second
-	hcDuration  = 500 * time.Millisecond
-)
+// Creates the network address for pushing service updates.
+func newPushAddr(host string) string {
+	return fmt.Sprintf("%s://%s:%s", pushProtocol, host, pushPort)
+}
+
+// Pushes service events upstream.
+func (p *Proxy) pushServiceEvent(meta Site, event ServiceEvent) error {
+	update := ServiceTableUpdate{
+		Meta:  meta,
+		Event: event,
+	}
+	jsn, err := json.Marshal(update)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest("POST", p.pushAddr, bytes.NewBuffer(jsn))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	client := &http.Client{}
+	go func() {
+		for {
+			resp, err := client.Do(req)
```

Comment:
if you wrap this in a `func` you can `defer resp.Body.Close()`. I was going to recommend putting the entire go routine instantiation in a loop. but I fear run away go routines. We should also properly track this go routine. if it dies or exits, we should restart.